File: module_6/java/CourseSchedule.java
/*
 *   Title: Course schedule
 *
 *   Problem:
 *   There are a total of n courses you have to take, labeled from 0 to n-1.

     Some courses may have prerequisites, for example to take course 0 you have to
     first take course 1, which is expressed as a pair: [0,1]

     Given the total number of courses and a list of prerequisite pairs, is it
     possible for you to finish all courses?
 *
 *   Execution: javac CourseSchedule.java && java CourseSchedule
 */
import java.util.*;


public class CourseSchedule {

    public static boolean canFinish(int n, int[][] a) {
        Map<Integer, List<Integer>> g = new HashMap();
        for(int[] i : a) {
            int k = i[1];
            List<Integer> cs = g.containsKey(k) ? g.get(k) : new ArrayList();
            cs.add(i[0]);
            g.put(k, cs);
        }

        Set<Integer> visited = new HashSet();
        for(int k : g.keySet()) {
            int c = helper(n, k, g, visited, new HashSet());
            if(c == -1) {
                return false;
            }

            if(c == n) {
                return true;
            }
        }

        return true;
    }

    public static int helper(int n, int k, Map<Integer, List<Integer>> g,
                       Set<Integer> visited, Set<Integer> visiting) {
        if(visiting.contains(k)) {
            return -1;
        }

        if(visited.contains(k)) {
            return 0;
        }

        int count = 1;

        if(g.containsKey(k)) {
            visiting.add(k);
            List<Integer> cs = g.get(k);
            for(int i : cs) {
                if(count >= n) {
                    return count;
                }
                int t = helper(n, i, g, visited, visiting);
                if(t < 0) {
                    return -1;
                }
                count += t;
            }
            visiting.remove(k);
        }

        visited.add(k);

        return count;
    }

    public static void main(String[] args) {
        int[][] test_input_1 = {{1, 0}};
        assert canFinish(2, test_input_1) == true;
        System.out.println("Explanation: There are a total of 2 courses to take. To take course 1 you should have finsihed course 0. So it is possible.");

        int[][] test_input_2 = {{1, 0}, {0, 1}};
        assert canFinish(2, test_input_2) == false;
        System.out.println("Explanation: There are  a total of 2 courses to take. To take course 1 you should have finished course 0 and to take course 0 you should also have finished course 1. So it is impossible");

        System.out.println("Passed all test cases");
    }
    
}

File: module_6/python/course_schedule.py
"""
Title: Course schedule

Problem:
    There are a total of n courses you have to take, labeled from 0 to n-1.

    Some courses may have prerequisites, for example to take course 0 you have to
    first take course 1, which is expressed as a pair: [0,1]

    Given the total number of courses and a list of prerequisite pairs, is it
    possible for you to finish all courses?

Execution: python course_scheule.py
"""
import unittest
from collections import defaultdict
from typing import List


def can_finish(num_courses: int, prerequisites: List[List[int]]) -> bool:
    """Determines if it is possible to finish all courses."""
    def DFS(start, my_dict, course_state):
        course_state[start] = 1
        for pre_course in my_dict[start]:
            if course_state[pre_course] == 1:
                return True
            if course_state[pre_course] == 0:
                if DFS(pre_course, my_dict, course_state):
                    return True
        course_state[start] = 2
        return False
    
    if not num_courses or not prerequisites:
        # Assume no course to take returns True
        return True
    
    my_dict = defaultdict(list)
    for p in prerequisites:
        my_dict[p[0]].append(p[1])

    # Init states for all courses
    course_state = [0] * num_courses

    for n in range(num_courses):
        if course_state[n] == 0: # Call DFS from this node and look for a loop
            loop = DFS(n, my_dict, course_state)
            if loop:
                return False
    return True  


class TestCourseSchedule(unittest.TestCase):
    """Unit test for can_finish for course schedule."""

    def test_1(self):
        self.assertEqual(can_finish(2, [[1, 0]]), True)
        print("Explanation: There are a total of 2 courses to take. To take course 1 you should have finished course 0. So it is possible.")

    def test_2(self):
        self.assertEqual(can_finish(2, [[1,0],[0,1]]), False)
        print("Explanation: There are a total of 2 courses to take. To take course 1 you should have finished course 0, and to take course 0 you should also have finished course 1. So it is impossible.")


if __name__ == '__main__':
    unittest.main()
